main/src/library/Vector.flix

```diff
@@ -0,0 +1,1389 @@
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+///
+/// A Vector data type which is fundamentally like Array but is immutable.
+///
+
+type alias Vector[a] = Array[a, true]
+
+namespace Vector {
+
+    ///
+    /// Compares `a` and `b` lexicographically.
+    ///
+    pub def compare(a: Vector[a], b: Vector[a]): Comparison with Order[a] =
+        let len = Int32.min(length(a), length(b));
+        def loop(i) = {
+            if (i < len) {
+                let cmp = get(i, a) <=> get(i, b);
+                if (cmp == Comparison.EqualTo)
+                    loop(i + 1)
+                else
+                    cmp
+            } else if (i < Array.length(a)) {
+                Comparison.GreaterThan
+            } else if (i < Array.length(b)) {
+                Comparison.LessThan
+            } else {
+                Comparison.EqualTo
+            }
+        };
+        loop(0)
+
+    ///
+    /// Returns a string representation of the given vector `v`.
+    ///
+    pub def toString(v: Vector[a]): String with ToString[a] = region rc {
+        let sb = new StringBuilder(rc);
+        let first = ref true @ rc;
+
+        StringBuilder.appendString!("Vector#{", sb);
+
+        let f = (x) -> {
+            if (not deref first) StringBuilder.appendString!(", ", sb) else first := false;
+            StringBuilder.appendString!("${x}", sb)
+        };
+        forEach(f, v);
+
+        StringBuilder.appendString!("}", sb);
+        StringBuilder.toString(sb)
+    }
+
+    ///
+    /// Returns an empty (length zero) vector.
+    ///
+    pub def empty(): Vector[a] = region rc {
+        buildVector(rc, 0, _ -> ())
+    }
+
+    ///
+    /// Returns a singleton vector containing `x``.
+    ///
+    pub def singleton(x: a): Vector[a] =
+        init(_ -> x, 1)
+
+    ///
+    /// Retrieves the value at position `i` in the vector `v`.
+    ///
+    pub def get(i: Int32, v: Vector[a]): a =
+        Array.get(i, v)
+
+    ///
+    /// Optionally returns the element at position `i` in the vector `v`.
+    ///
+    pub def nth(i: Int32, v: Vector[a]): Option[a] =
+        if (0 <= i and i < length(v))
+            Some(get(i, v))
+        else
+            None
+
+    ///
+    /// Returns `true` if the given vector `v` is empty.
+    ///
+    pub def isEmpty(v: Vector[a]): Bool = length(v) == 0
+
+    ///
+    /// Returns the length of the vector `v`.
+    ///
+    pub def length(v: Vector[a]): Int32 = Array.length(v)
+
+    ///
+    /// Returns a fresh vector.
+    ///
+    /// The function `f` is applied to a uninitialized array, it should should set each element
+    /// to an actual value.
+    ///
+    pub def buildVector(rc: Region[r], len: Int32, f: Array[a, r] -> Unit \ ef): Vector[a] \ ef =
```

Comment:
So if this becomes a primitive, we should be good?

Reply:
I'm in the process of removing `buildVector` - it's doing nothing that a `region` block doesn't do. In a language without regions it would be useful to encapsulate the new array, but that's exactly what regions do so it can be eliminated in Flix.